CPU model, topology and NUMA node tuning  GSoC

gqlo opened the pull request:
### What does this PR do?
In response to feature request #57880 

This PR add the feature to support CPU model and topology. configurable parameters are:
- `mode`
- `match`
- `check`
- `model`
- `vendor`
- `topology`
- `cache` 
- `feature` 

`numa` which is used to specify the attributes and sub-elements of numa `guest`. configurable parameters are:
- `cpus`
- `memory`
- `memAccess`
- `distance`
- `sibling` id and values

To configure numa `guest`,  numa `host ` must be configured beforehand, `numatune` can be used to specify `memory` which tells how to allocate memory for the domain process on a NUMA host and `memnode` which specify memory allocation policies per each guest NUMA node.

See [Libvirt domain XML](https://libvirt.org/formatdomain.html#elementsCPU) for more details.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltstack.com/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltstack.com/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltstack.com/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

cbosdo commented on salt/modules/virt.py:
```diff
@@ -740,35 +740,48 @@ def _gen_xml(
+                                    context["cpu"]["numa"][numa_id].update(
+                                        {k: "{}='{}'".format(k, v)}
+                                    )
+        print("----------------context after ---------------------", context["cpu"])
```
another debug print to be removed

cbosdo commented on salt/modules/virt.py:
```diff
@@ -1726,7 +1739,66 @@ def init(
+                 1:
+                   enabled: 'no'
+                   hotpluggable: 'yes'
+               attrib:
+                 match: minimum
+                 mode: custom
+                 check: full
```
I would move `match`, `mode` and `check` one level up to remove `attrib`

gqlo replied:
got an idea already :)

cbosdo commented on salt/modules/virt.py:
```diff
@@ -2164,6 +2296,7 @@ def seeder(path):
             numatune,
             **kwargs
         )
+        print("---------------------vm_xml-------------------------", vm_xml)
```
debug print to be removed or converted to `log.debug()`

cbosdo reviewed (changes requested):
First round of comments. Code looks nice, may be simplified / factorized in some places and more important group commits to avoid the same code to change in every commit: makes the whole thing hard to read ;)

gqlo commented on salt/modules/virt.py:
```diff
@@ -735,10 +736,58 @@ def _gen_xml(
+                                )
+                            }
+                        )
+            elif key in ["attrib", "model", "topology", "cache", "feature"]:
```
It's possible to get rid of `attrib`  But it might make the jinja code look even more ugly and less readable. This is how I handle `cpu attribute` in jinja templates  `<cpu {% for attr in cpu.attrib %} {{ cpu.attrib[attr] + ''}} {%- endfor %}>` There are 3 possible cpu attributes and 2 of them are optional, I can't think of a neater way of handle it. Do you have better suggestions?

gqlo replied:
Actually, i just had an idea of get rid of attrib without changing the jinja template :)

bryceml commented:
We should make sure the added options make it into the documentation.

gqlo commented:
> We should make sure the added options make it into the documentation with example configuration.
@bryceml Hi Thanks for the comments. yeah i need to work on the documentation a bit more.

cbosdo reviewed (changes requested):
Refactoring in progress to make the code more generic. Do not merge as is

cbosdo commented on salt/modules/virt.py:
```diff
@@ -2757,10 +3179,8 @@ def _append_placeholder(value_dict, key):
                     for placeholder in placeholders
                 }
                 node_xpath = string.Template(xpath).substitute(ctx)
-
```
please avoid removing blank lines

gqlo committed:
virt: support host numa tunning capability

fixup! virt: support host numa tunning capability

gqlo committed:
virt: support cpu model and topology

fixup! virt: support cpu model and topology

cbosdo commented on tests/unit/modules/test_virt.py:
```diff
@@ -2715,6 +2819,176 @@ def test_update_existing_boot_params(self):
+            setxml.find("./numatune/memnode/[@cellid='4']").attrib.get("nodeset"), "6"
+        )
+
+        self.assertEqual(setxml.find("./numatune/memnode/[@cellid='2']"), None)
```
Why checking for a `memnode` that was never in the original nor updated description? better ensure that `cellid='1'` has been removed.

gqlo replied:
I will add the assertion to ensure cellid='1' is removed. I guess it's still a good practice to check `memnode` that was never updated. Just in case there is something wrong with the update logic. We could run into the case that trying to update `cellid=1` which might messed with `cellid=2`. But i agree this is trivial.

cbosdo commented on tests/unit/modules/test_virt.py:
```diff
@@ -2715,6 +2819,176 @@ def test_update_existing_boot_params(self):
+            setxml.find("./numatune/memnode/[@cellid='3']").attrib.get("nodeset"), "7"
+        )
+
+        self.assertEqual(setxml.find("./numatune/memnode/[@cellid='2']"), None)
```
This assert should be deleted too

cbosdo commented on tests/unit/modules/test_virt.py:
```diff
@@ -2715,6 +2819,176 @@ def test_update_existing_boot_params(self):
+
+        self.assertEqual(setxml.find("./numatune/memnode/[@cellid='2']"), None)
+
+        numatune_mnodes_none = {
+            "memory": {"mode": "preferred", "nodeset": "0-5"},
+            "memnodes": None,
+        }
```
Instead of this test and the previous one, you should merge them into one that removes all numatune settings. It surely covers the same cases and would make it shorter

cbosdo commented on tests/unit/modules/test_virt.py:
```diff
@@ -2715,6 +2819,176 @@ def test_update_existing_boot_params(self):
+            },
+        }
+
+        self.assertEqual(
+            {
+                "definition": True,
+                "disk": {"attached": [], "detached": [], "updated": []},
+                "interface": {"attached": [], "detached": []},
+            },
+            virt.update("vm_with_existing_param", numatune=numatune_mnode_none),
+        )
+
+        setxml = ET.fromstring(define_mock.call_args[0][0])
+        self.assertEqual(
+            setxml.find("numatune").find("memory").attrib.get("mode"), "preferred"
+        )
+
+        self.assertEqual(
+            setxml.find("numatune").find("memory").attrib.get("nodeset"), "0-5"
+        )
+
+        self.assertEqual(setxml.find("./numatune/memnode/[@cellid='0']"), None)
+
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='3']").attrib.get("mode"),
+            "preferred",
+        )
+
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='3']").attrib.get("nodeset"), "7"
+        )
+
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='4']").attrib.get("mode"), "strict"
+        )
+
+        self.assertEqual(
+            setxml.find("./numatune/memnode/[@cellid='4']").attrib.get("nodeset"), "6"
+        )
+
```
I would remove that test since it's not that useful, and replace it by a test with exactly the parameters of the existing definition to ensure we're not updating the VM in such cases.

cbosdo commented on salt/modules/virt.py:
```diff
@@ -883,6 +882,76 @@ def _gen_xml(
+                                            )
+                                        else:
+                                            context["cpu"]["vcpus"][vcpu_id][
+                                                k
+                                            ] = "{}='{}'".format(k, v)
+                    elif tag in ["model", "topology", "cache", "feature"]:
+                        for k, v in cpu.get(tag).items():
+                            if not (tag == "model" and k == "name"):
+                                if v:
+                                    context["cpu"][tag][k] = "{}='{}'".format(k, v)
+                            else:
+                                if v:
+                                    context["cpu"][tag][k] = v
+                    elif tag == "numa":
+                        numa_param = cpu.get(tag)
+                        for cell_id in numa_param:
+                            if numa_param.get(cell_id):
+                                for k, v in numa_param[cell_id].items():
+                                    if k == "memory":
+                                        if v:
+                                            context["cpu"]["numa"][cell_id][
+                                                k
+                                            ] = "{}='{}'".format(
+                                                k, int(_handle_unit(v) / 1024)
+                                            )
+                                    elif k == "distances":
+                                        if v:
+                                            context["cpu"]["numa"][cell_id][k] = v
+                                    else:
+                                        if v is not None:
+                                            if k == "discard":
+                                                context["cpu"]["numa"][cell_id][
+                                                    k
+                                                ] = "{}='{}'".format(
+                                                    k, "yes" if v else "no"
+                                                )
+                                            else:
+                                                context["cpu"]["numa"][cell_id][
+                                                    k
+                                                ] = "{}='{}'".format(k, v)
```
I thought you implemented something simpler than that here... has it slipped in another PR?

cbosdo commented on salt/modules/virt.py:
```diff
@@ -1955,6 +2081,70 @@ def init(
+    The cpu parameters dictionary can contain the following properties:
+
+    cpuset
+        a comma-separated list of physical CPU numbers that domain process and virtual CPUs can be pinned to by default
```
May be you should provide examples here. It also can handle ranges and excluded CPUs like `1-4,^3,6`

cbosdo commented on salt/modules/virt.py:
```diff
@@ -1955,6 +2081,70 @@ def init(
+    current
+        the number of virtual cpus available at startup
+
+    placement
+        indicate the CPU placement mode for domain process. the value can be either ``static`` or ``auto``
```
Shouldn't we make it easier by transforming into `auto-placement` as a boolean value defaulting to `True`?

cbosdo commented on salt/modules/virt.py:
```diff
@@ -1955,6 +2081,70 @@ def init(
+        ``mode`` attribute supported three possible values: ``emulate``, ``passthrough``, ``disable``
+
+    numa
+        specify the guest numa topology. ``cell`` element specifies a NUMA cell or a NUMA node, ``cpus`` specifies the
+        CPU or range of CPUs that are part of the node, ``memory`` specifies the node memory in kibibytes. All cells
+        should have ``id`` attribute in case referring to some cell is necessary in the code. optional attribute
+        ``memAccess`` control whether the memory is to be mapped as ``shared`` or ``private``, ``discard`` attribute which
+        fine tunes the discard feature for given numa node, possible values are ``yes`` or ``no``.  ``distances`` element
+        define the distance between NUMA cells and ``sibling`` sub-element is used to specify the distance value between
+        sibling NUMA cells.
```
Seems that documentation fragment is outdated already ;)

cbosdo commented on salt/modules/virt.py:
```diff
@@ -2528,7 +2718,7 @@ def _diff_graphics_lists(old, new):
 
 def update(
     name,
-    cpu=0,
+    cpu=None,
```
Do we really have to change the default value?

cbosdo commented on salt/modules/virt.py:
```diff
@@ -2809,6 +3006,184 @@ def _get_with_unit(node):
+            "set": _set_vcpu,
+        },
+        {"path": "cpu:maximum", "xpath": "vcpu", "get": lambda n: int(n.text)},
+        {
+            "path": "cpu:placement",
+            "xpath": "vcpu",
+            "get": lambda n: n.get("placement"),
+            "set": lambda n, v: n.set("placement", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("placement"),
+        },
```
If converting to `auto-placement` then this should be changed too

gqlo replied:
It seems like we can pass `placement:auto` to indicate auto placement

cbosdo replied:
Sure, but the idea here was to use a boolean rather than a string to avoid the user to look for the possible values in the doc.

cbosdo commented on salt/modules/virt.py:
```diff
@@ -2809,6 +3006,184 @@ def _get_with_unit(node):
+            "set": lambda n, v: n.set("mode", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("mode"),
+        },
+        {
+            "path": "cpu:feature:name",
+            "xpath": "cpu/feature",
+            "get": lambda n: n.get("name"),
+            "set": lambda n, v: n.set("name", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("name"),
+        },
+        {
+            "path": "cpu:feature:policy",
+            "xpath": "cpu/feature",
+            "get": lambda n: n.get("policy"),
+            "set": lambda n, v: n.set("policy", str(v)),
+            "del": salt.utils.xmlutil.del_attribute("policy"),
+        },
```
There is an issue here: there can (are likely to) be several features. We need to handle things like:
```
cpu:
  features:
    vmx: require
    lahf_lm: disable
```

cbosdo commented on tests/unit/modules/test_virt.py:
```diff
@@ -1841,13 +1841,282 @@ def test_update(self):
+        self.assertEqual(setxml.find("cpu").find("cache").attrib.get("mode"), "emulate")
+
+        # test adding feature
+        cpu_feature = {"feature": {"policy": "require", "name": "pcid"}}
```
Please test adding multiple features instead.

cbosdo commented on tests/unit/modules/test_virt.py:
```diff
@@ -1841,13 +1841,282 @@ def test_update(self):
             "cmdline": "console=ttyS0 ks=http://example.com/f8-i386/os/",
         }
 
+        boot_uefi = {
+            "loader": "/usr/share/OVMF/OVMF_CODE.fd",
+            "nvram": "/usr/share/OVMF/OVMF_VARS.ms.fd",
+        }
+
+        invalid_boot = {
+            "loader": "/usr/share/OVMF/OVMF_CODE.fd",
+            "initrd": "/root/f8-i386-initrd",
+        }
+
```
Hum, this is surely a rebase issue ;)

cbosdo reviewed (changes requested):
There are still a few important things to fix in this PR, see the inline comments

gqlo commented on salt/modules/virt.py:
```diff
@@ -1937,6 +1940,21 @@ def init(
+        The optional numatune element provides details of how to tune the performance of a NUMA host via controlling NUMA
+        policy for domain process. The optional ``memory`` element specifies how to allocate memory for the domain process
+        on a NUMA host. ``memnode`` elements can specify memory allocation policies per each guest NUMA node. The definition
+        used in the dictionary can be found at :ref:`init-cpu-def`.
```
hmm, actually `nodeset` `mode` are coming from `init-cpu-def`. They are somehow connected.

cbosdo replied:
I am still convinced that the `numatune` parameter description shouldn't be shared with the cpu one, just for the sake of clarity. Otherwise you need to have clear references to the `init-cpu-def` sub section you mention. This starts to grow big, let's make it easy for the users.

gqlo commented on salt/states/virt.py:
```diff
@@ -286,6 +286,7 @@ def defined(
     os_type=None,
     arch=None,
     boot=None,
+    numatune=None,
```
thanks for pointing that out. I will add them

s0undt3ch commented on tests/unit/modules/test_virt.py:
```diff
@@ -1871,7 +1871,7 @@ def test_update(self):
         setxml = ET.fromstring(define_mock.call_args[0][0])
         self.assertEqual("2", setxml.find("vcpu").text)
-        self.assertEqual("2048", setxml.find("memory").text)
+        self.assertEqual("2147483648", setxml.find("memory").text)
```
Can you please submit this fix in a separate PR.
I'd like to merge that sooner than this PR.
Thanks!

Commit:
fixup! precommit failure fix

gqlo commented:
re-run all

gqlo commented:
re-run all

Ch3LL commented on salt/modules/virt.py:
```diff
@@ -2578,6 +2791,7 @@ def update(
     graphics=None,
     live=True,
     boot=None,
+    numatune=None,
```
this kwarg should be after `boot_dev` to safely cover the case for someone who is calling these kwargs as positional parameters.

gqlo replied:
That's right ! Thanks Megan.

cbosdo committed:
virt: make context preprocessing more reusable in _gen_xml

Introduce mapping structures in order to help reusing the common patterns
in the virt._gen_xml() context pre processing.

cbosdo reviewed (changes requested):
Still some doc changes to be made. Also look at https://github.com/cbosdo/salt/commit/90b6e0c8339b29d7c41f6502f2e911029095bf41 for a suggestion of how to better handle the `_gen_xml` added code.

cbosdo committed:
libvirt domain template memory config fixes

Add unit tests for _gen_xml() on the recently added memory parameters.
Also fixes an issue with an optional attribute.

gqlo commented:
Commits of this PR is merged by another one.
